experiments/2017-09-21-rust-on-android.md

```diff
@@ -0,0 +1,224 @@
+    }
+}
+```
+The first line here `#[cfg(target_os="android")]` is telling the compiler to target Android when compiling this module. `#[cfg]` is a special attribute that allows you to compile code based on a flag passed to the compiler. 
+
+The second line, `#[allow(non_snake_case)]`, tells the compiler not to warn if we are not using snake_case for a variable or function name. The Rust compiler is very strict, this is one of the things that makes Rust great - it's nearly impossible to code something that is not correct, and it enforces the use of snake_case throughout. However, we defined our class name and native method in our Android project using Java coding conventions which is camelCase and UpperCamelCase and we don't want to change this or our Java code will look wrong. Given the way that JNI constructs native function names, we need to tell the Rust compiler to go easy on us in this instance. This flag will apply to all functions and variables created inside this module that we are creating, called `android`. 
+
+After declaring that we need the `jni` create and importing some things from it, we can declare our function. This function needs to be marked `unsafe` because we will be dealing with pointers from a language that allows null pointers, but our code doesn't check for `NULL`. This situation would never happen in strictly Rust code as the Rust compiler enforces memory safety. By marking the function as not memory safe, we are alerting other Rust functions that it may not be able to deal with a null pointer. `extern` defines the function as one that will be exposed to other languages. 
+
+As arguments, along with the `JString` that our Java function declaration said that we will be providing, we also need to take an instance of the `JNIEnv` and a class reference. The `JNIEnv` will be the object we will use to read values associated with the pointers that we are taking as argument.
+
+Next, we read the string in from the `JNIEnv` and convert it into a C pointer to pass to `rust_greeting`. The result of that function is another C pointer, which we then need to convert to a back into a String. Using the `JNIEnv` transfers the ownership of the object to Java, but there is still a reference hanging around held by our Rust code. We need to call `rust_greetings_free` to free that memory. Then we return our String.
+
+We declared that we needed the `jni` crate, that means we need to include the crate in the `Cargo.toml` file. Open it up and add the following between the `[package]` and `[lib]` declarations.
+```
+[target.'cfg(target_os="android")'.dependencies]
+jni = { version = "0.5", default-features = false }
+```
+
+We are now ready to build our libraries. Unlike with iOS, there is no handy universal Android library that we can make so we have to create one for each of our target architectures. We can then create symlinks to them from the Android project.
+```
+cargo build --target aarch64-linux-android --release
```

Comment:
I think we're in the `cargo` dir here (or whatever we're calling it)? We've diverted into setting up a Android Studio project since we were last there.

I got an error at this point:

```
$ cargo build --target aarch64-linux-android --release --verbose
       Fresh byteorder v1.1.0
       Fresh error-chain v0.10.0
       Fresh ascii v0.7.1
       Fresh jni-sys v0.2.5
       Fresh cesu8 v1.1.0
       Fresh log v0.3.8
       Fresh combine v2.5.1
       Fresh jni v0.5.3
   Compiling greetings v0.1.1 (file:///Users/joe/projects/mozilla/greetings/cargo)
     Running `rustc --crate-name greetings src/lib.rs --crate-type staticlib --crate-type cdylib --emit=dep-info,link -C opt-level=3 -C metadata=0cf6bfc9208123c1 --out-dir /Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps --target aarch64-linux-android -L dependency=/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps -L dependency=/Users/joe/projects/mozilla/greetings/cargo/target/release/deps --extern jni=/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/libjni-cb346174c42792ff.rlib`
note: link against the following native artifacts when linking against this static library

note: the order and any duplication can be significant on some platforms, and so may need to be preserved

note: library: dl

note: library: log

note: library: gcc

note: library: c

note: library: m

error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-Wl,--allow-multiple-definition" "-L" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/greetings.0.o" "-o" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/libgreetings.so" "-Wl,--version-script=/var/folders/hm/2lmq6xk10hg0d0f9k_s51w4w0000gn/T/rustc.4gGG1cEbIsCs/list" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/greetings.crate.allocator.o" "-Wl,--gc-sections" "-Wl,-O1" "-nodefaultlibs" "-L" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps" "-L" "/Users/joe/projects/mozilla/greetings/cargo/target/release/deps" "-L" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib" "-Wl,-Bstatic" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/libjni-cb346174c42792ff.rlib" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/liberror_chain-a73b095fe0c717fb.rlib" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/libjni_sys-48e34f27f88e1b9d.rlib" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/libcombine-df5b4f5a4cbbf4e3.rlib" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/libascii-4af0da923a5eef36.rlib" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/libcesu8-0462512f8e378108.rlib" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/liblog-68979984efb14d74.rlib" "/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/libbyteorder-1f8143e2002fa3c6.rlib" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib/libstd-881b9fdbdf1d515b.rlib" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib/libpanic_unwind-9cbadc6554202be9.rlib" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib/libunwind-618c266cf9124966.rlib" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib/liballoc_system-3c10208cdd7e61cb.rlib" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib/liblibc-16f3b02b9a976b94.rlib" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib/liballoc-bc70b4efeaeb398c.rlib" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib/libstd_unicode-a0ad42dc8f5856aa.rlib" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib/librand-32f648f7f7567c6c.rlib" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib/libcore-21491ce3d14f1ef2.rlib" "/Users/joe/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/aarch64-linux-android/lib/libcompiler_builtins-b9713bd7f605c0b2.rlib" "-Wl,-Bdynamic" "-l" "dl" "-l" "log" "-l" "gcc" "-l" "gcc" "-l" "c" "-l" "m" "-shared"
  = note: ld: unknown option: --as-needed
          clang: error: linker command failed with exit code 1 (use -v to see invocation)


error: aborting due to previous error

error: Could not compile `greetings`.

Caused by:
  process didn't exit successfully: `rustc --crate-name greetings src/lib.rs --crate-type staticlib --crate-type cdylib --emit=dep-info,link -C opt-level=3 -C metadata=0cf6bfc9208123c1 --out-dir /Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps --target aarch64-linux-android -L dependency=/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps -L dependency=/Users/joe/projects/mozilla/greetings/cargo/target/release/deps --extern jni=/Users/joe/projects/mozilla/greetings/cargo/target/aarch64-linux-android/release/deps/libjni-cb346174c42792ff.rlib` (exit code: 101)
```

Reply:
Did you do `cp cargo-config.toml ~/.cargo/config` or leave your `cargo-config.toml` in the root directory? It looks as if it's not finding the standalone NDK for your architecture.